src/theory/quantifiers/sygus/sygus_grammar_red.cpp

```diff
@@ -29,150 +26,148 @@ namespace cvc5::internal {
-}
-
-void SygusRedundantCons::getRedundant(std::vector<unsigned>& indices)
-{
-  const DType& dt = d_type.getDType();
-  for (unsigned i = 0, ncons = dt.getNumConstructors(); i < ncons; i++)
+  if (TraceIsOn("sygus-grammar-red"))
   {
-    if (isRedundant(i))
+    Trace("sygus-grammar-red") << "Variants of " << r << ":" << std::endl;
+    for (const Node& t : tset)
     {
-      indices.push_back(i);
+      Trace("sygus-grammar-red") << "  " << t << std::endl;
     }
   }
+  return tset;
 }
 
-bool SygusRedundantCons::isRedundant(unsigned i)
-{
-  Assert(i < d_sygus_red_status.size());
-  return d_sygus_red_status[i] == 1;
-}
-
-void SygusRedundantCons::getGenericList(TermDbSygus* tds,
-                                        const DType& dt,
-                                        unsigned c,
-                                        unsigned index,
-                                        std::map<int, Node>& pre,
-                                        std::vector<Node>& terms)
+void SygusGrammarReduce::getGenericListRec(
+    const Node& lam,
+    std::unordered_set<Node>& tset,
+    const std::vector<std::pair<Node, size_t>>& vlist,
+    const std::vector<Node>& ntlist,
+    std::map<Node, std::vector<Node>>& ntvMap,
+    size_t ntindex,
+    size_t vindex)
 {
-  if (index == dt[c].getNumArgs())
+  // to avoid exponential behavior, stop if we have more than 10 variants
+  // already
+  if (tset.size() >= 10)
   {
-    Node gt = tds->mkGeneric(dt, c, pre);
-    gt = extendedRewrite(gt);
-    terms.push_back(gt);
     return;
   }
-  // with no swap
-  getGenericList(tds, dt, c, index + 1, pre, terms);
-  // swapping is exponential, only use for operators with small # args.
-  if (dt[c].getNumArgs() <= 5)
+  if (ntindex == ntlist.size())
   {
-    TypeNode atype = tds->getArgType(dt[c], index);
-    for (unsigned s = index + 1, nargs = dt[c].getNumArgs(); s < nargs; s++)
+    std::vector<Node> args;
+    args.push_back(lam);
+    for (const std::pair<Node, size_t>& v : vlist)
     {
-      if (tds->getArgType(dt[c], s) == atype)
-      {
-        // swap s and index
-        Node tmp = pre[s];
-        pre[s] = pre[index];
-        pre[index] = tmp;
-        getGenericList(tds, dt, c, index + 1, pre, terms);
-        // revert
-        tmp = pre[s];
-        pre[s] = pre[index];
-        pre[index] = tmp;
-      }
+      args.push_back(ntvMap[v.first][v.second]);
     }
+    Node app = NodeManager::currentNM()->mkNode(Kind::APPLY_UF, args);
+    // apply extended rewriting to maximize chances for duplication
+    app = extendedRewrite(app);
+    tset.insert(app);
+    return;
+  }
+  Node nts = ntlist[ntindex];
+  std::vector<Node>& ntvs = ntvMap[nts];
+  if (vindex == ntvs.size())
+  {
+    // go to next non-terminal
+    return getGenericListRec(lam, tset, vlist, ntlist, ntvMap, ntindex + 1, 0);
+  }
+  for (size_t i = vindex, nvars = ntvs.size(); i < nvars; i++)
+  {
+    // swap the variables
+    std::swap(ntvs[i], ntvs[vindex]);
```

Comment:
It's not clear to me if this simulates all permutations. It would be simpler to just return a list of all permutations and then iterate through the list.

Reply:
The problem is that the number of permutations is exponential. This method bounds the number of considered permutations.